File: aforp/lib/models/equipment.dart
class Equipment {
  final int? id;
  final String name;
  final String description;
  final String? imageUrl;
  final int totalQuantity;
  final int availableQuantity;
  final String? category;
  final String condition;
  final bool isAvailable;
  final DateTime? createdAt;
  final DateTime? updatedAt;

  Equipment({
    this.id,
    required this.name,
    required this.description,
    this.imageUrl,
    required this.totalQuantity,
    required this.availableQuantity,
    this.category,
    required this.condition,
    bool? isAvailable,
    this.createdAt,
    this.updatedAt,
  }) : isAvailable = isAvailable ?? (availableQuantity > 0);

  factory Equipment.fromJson(Map<String, dynamic> json) {
    final availableQty = json['available_quantity'] is String 
        ? int.tryParse(json['available_quantity']) ?? 0 
        : json['available_quantity'] ?? 0;
        
    return Equipment(
      id: json['id'],
      name: json['name'] ?? '',
      description: json['description'] ?? '',
      imageUrl: json['image_url'],
      totalQuantity: json['total_quantity'] is String 
          ? int.tryParse(json['total_quantity']) ?? 0 
          : json['total_quantity'] ?? 0,
      availableQuantity: availableQty,
      category: json['category'],
      condition: json['condition'] ?? 'Bon',
      isAvailable: json['is_available'] != null ? 
          (json['is_available'] == 1 || json['is_available'] == true) : 
          (availableQty > 0),
      createdAt: json['created_at'] != null
          ? DateTime.parse(json['created_at'])
          : null,
      updatedAt: json['updated_at'] != null
          ? DateTime.parse(json['updated_at'])
          : null,
    );
  }

  Map<String, dynamic> toJson() {
    return {
      if (id != null) 'id': id,
      'name': name,
      'description': description,
      'image_url': imageUrl,
      'total_quantity': totalQuantity,
      'available_quantity': availableQuantity,
      'category': category,
      'condition': condition,
      'is_available': isAvailable ? 1 : 0,
      if (createdAt != null) 'created_at': createdAt?.toIso8601String(),
      if (updatedAt != null) 'updated_at': updatedAt?.toIso8601String(),
    };
  }

  Equipment copyWith({
    int? id,
    String? name,
    String? description,
    String? imageUrl,
    int? totalQuantity,
    int? availableQuantity,
    String? category,
    String? condition,
    bool? isAvailable,
    DateTime? createdAt,
    DateTime? updatedAt,
  }) {
    final newAvailableQuantity = availableQuantity ?? this.availableQuantity;
    
    return Equipment(
      id: id ?? this.id,
      name: name ?? this.name,
      description: description ?? this.description,
      imageUrl: imageUrl ?? this.imageUrl,
      totalQuantity: totalQuantity ?? this.totalQuantity,
      availableQuantity: newAvailableQuantity,
      category: category ?? this.category,
      condition: condition ?? this.condition,
      isAvailable: isAvailable ?? (newAvailableQuantity > 0),
      createdAt: createdAt ?? this.createdAt,
      updatedAt: updatedAt ?? this.updatedAt,
    );
  }
} 

File: backend/routes/equipment.js
const express = require('express');
const router = express.Router();
const jwt = require('jsonwebtoken');
const pool = require('../config/db');

// Middleware pour vérifier le token JWT
const verifyToken = (req, res, next) => {
  const authHeader = req.headers.authorization;
  
  if (!authHeader || !authHeader.startsWith('Bearer ')) {
    return res.status(401).json({ message: 'Token non fourni' });
  }
  
  const token = authHeader.split(' ')[1];
  
  try {
    const decoded = jwt.verify(token, process.env.JWT_SECRET);
    req.user = decoded;
    next();
  } catch (error) {
    console.error('Erreur de vérification du token:', error);
    return res.status(401).json({ message: 'Token invalide' });
  }
};

// Middleware pour vérifier si l'utilisateur est admin
const verifyAdmin = async (req, res, next) => {
  try {
    const [users] = await pool.query('SELECT role FROM users WHERE id = ?', [req.user.id]);
    
    if (users.length === 0) {
      return res.status(404).json({ message: 'Utilisateur non trouvé' });
    }
    
    if (users[0].role !== 'admin') {
      return res.status(403).json({ message: 'Accès refusé: rôle admin requis' });
    }
    
    next();
  } catch (error) {
    console.error('Erreur lors de la vérification du rôle admin:', error);
    res.status(500).json({ message: 'Erreur serveur: ' + error.message });
  }
};

// Route pour récupérer tous les équipements
router.get('/', verifyToken, async (req, res) => {
  console.log('Route GET /equipment appelée par l\'utilisateur ID:', req.user.id);
  try {
    const [equipment] = await pool.query('SELECT * FROM equipment');
    
    console.log(`${equipment.length} équipements récupérés`);
    res.json(equipment);
  } catch (error) {
    console.error('Erreur lors de la récupération des équipements:', error);
    res.status(500).json({ message: 'Erreur serveur: ' + error.message });
  }
});

// Route pour récupérer uniquement les équipements disponibles
router.get('/available', verifyToken, async (req, res) => {
  console.log('Route GET /equipment/available appelée par l\'utilisateur ID:', req.user.id);
  try {
    const [equipment] = await pool.query('SELECT * FROM equipment WHERE available_quantity > 0');
    
    console.log(`${equipment.length} équipements disponibles récupérés`);
    res.json(equipment);
  } catch (error) {
    console.error('Erreur lors de la récupération des équipements disponibles:', error);
    res.status(500).json({ message: 'Erreur serveur: ' + error.message });
  }
});

// Route pour récupérer un équipement par son ID
router.get('/:id', verifyToken, async (req, res) => {
  const equipmentId = req.params.id;
  console.log(`Route GET /equipment/${equipmentId} appelée par l'utilisateur ID: ${req.user.id}`);
  
  try {
    const [equipment] = await pool.query('SELECT * FROM equipment WHERE id = ?', [equipmentId]);
    
    if (equipment.length === 0) {
      return res.status(404).json({ message: 'Équipement non trouvé' });
    }
    
    res.json(equipment[0]);
  } catch (error) {
    console.error(`Erreur lors de la récupération de l'équipement ID ${equipmentId}:`, error);
    res.status(500).json({ message: 'Erreur serveur: ' + error.message });
  }
});

// Route pour créer un nouvel équipement (admin seulement)
router.post('/', verifyToken, verifyAdmin, async (req, res) => {
  console.log('Route POST /equipment appelée avec:', req.body);
  try {
    const { 
      name, 
      description, 
      image_url,
      total_quantity, 
      available_quantity, 
      category,
      condition = 'Bon'
    } = req.body;
    
    // Insérer l'équipement dans la base de données
    const [result] = await pool.query(
      'INSERT INTO equipment (name, description, image_url, total_quantity, available_quantity, category, `condition`) VALUES (?, ?, ?, ?, ?, ?, ?)',
      [name, description, image_url, total_quantity, available_quantity || total_quantity, category, condition]
    );
    
    res.status(201).json({
      id: result.insertId,
      name,
      description,
      image_url,
      total_quantity,
      available_quantity: available_quantity || total_quantity,
      category,
      condition,
      created_at: new Date().toISOString(),
      updated_at: new Date().toISOString()
    });
  } catch (error) {
    console.error('Erreur lors de la création de l\'équipement:', error);
    res.status(500).json({ message: 'Erreur serveur: ' + error.message });
  }
});

// Route pour mettre à jour un équipement (admin seulement)
router.put('/:id', verifyToken, verifyAdmin, async (req, res) => {
  const equipmentId = req.params.id;
  console.log(`Route PUT /equipment/${equipmentId} appelée avec:`, req.body);
  
  try {
    const { 
      name, 
      description, 
      image_url,
      total_quantity, 
      available_quantity, 
      category,
      condition = 'Bon'
    } = req.body;
    
    // Vérifier si l'équipement existe
    const [equipment] = await pool.query('SELECT * FROM equipment WHERE id = ?', [equipmentId]);
    
    if (equipment.length === 0) {
      return res.status(404).json({ message: 'Équipement non trouvé' });
    }
    
    // Mettre à jour l'équipement
    await pool.query(
      'UPDATE equipment SET name = ?, description = ?, image_url = ?, total_quantity = ?, available_quantity = ?, category = ?, `condition` = ? WHERE id = ?',
      [name, description, image_url, total_quantity, available_quantity, category, condition, equipmentId]
    );
    
    res.json({
      id: parseInt(equipmentId),
      name,
      description,
      image_url,
      total_quantity,
      available_quantity,
      category,
      condition,
      updated_at: new Date().toISOString()
    });
  } catch (error) {
    console.error(`Erreur lors de la mise à jour de l'équipement ID ${equipmentId}:`, error);
    res.status(500).json({ message: 'Erreur serveur: ' + error.message });
  }
});

// Route pour supprimer un équipement (admin seulement)
router.delete('/:id', verifyToken, verifyAdmin, async (req, res) => {
  const equipmentId = req.params.id;
  console.log(`Route DELETE /equipment/${equipmentId} appelée`);
  
  try {
    // Vérifier si l'équipement existe
    const [equipment] = await pool.query('SELECT * FROM equipment WHERE id = ?', [equipmentId]);
    
    if (equipment.length === 0) {
      return res.status(404).json({ message: 'Équipement non trouvé' });
    }
    
    // Supprimer l'équipement
    await pool.query('DELETE FROM equipment WHERE id = ?', [equipmentId]);
    
    res.json({ message: 'Équipement supprimé avec succès' });
  } catch (error) {
    console.error(`Erreur lors de la suppression de l'équipement ID ${equipmentId}:`, error);
    res.status(500).json({ message: 'Erreur serveur: ' + error.message });
  }
});

module.exports = router; 